File: c++ version/1-100/79-word-search.cpp
vector<int> directions{-1, 0, 1, 0, -1};

bool exist(vector<vector<char>>& board, string word) {
    int m = board.size(), n = board[0].size();
    vector<vector<bool>> visited(m, vector<bool>(n, false));
    int wordPos = 0;
    bool find = false;
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            backtracking(i, j, board, word, visited, find, wordPos);
            if (find) return true;
        }
    }
    return false;
}

void backtracking(int i, int j, vector<vector<char>> &board, string &word,
    vector<vector<bool>> &visited, bool &find, int wordPos) {
    if (find || board[i][j] != word[wordPos])
        return;
    
    if (wordPos == word.size() - 1) {
        find = true;
        return;
    }

    visited[i][j] = true;
    for (int k = 0; k < 4; ++k) {
        int row = i + directions[k];
        int col = j + directions[k+1];
        if (row >= 0 && row < board.size() && col >= 0 && col < board[0].size() && 
            !visited[row][col]) {
            backtracking(row, col, board, word, visited, find, wordPos + 1);
        }
    }
    visited[i][j] = false;

}

File: python version/79-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        direction = [-1, 0, 1, 0, -1]
        n = len(board)
        m = len(board[0])
        self.visited = [[False for _ in range(m)] for _ in range(n)]
        self.found = False

        def backtracking(i, j, curr):
            if board[i][j] != word[curr] or self.found: return
            if curr == len(word)-1:
                self.found = True 
                return
            self.visited[i][j] = True
            for k in range(4):
                row = i + direction[k]
                col = j + direction[k+1]
                if row >= 0 and row <= n-1 and col >= 0 and col <= m-1 and not self.visited[row][col]:
                    backtracking(row, col, curr+1)
            self.visited[i][j] = False

        for i in range(n):
            for j in range(m):
                if not self.found:
                    backtracking(i, j, 0)
        
        return self.found
